src/pow.rs

```diff
@@ -0,0 +1,1314 @@
+        let shift = shift as usize;
+
+        let Uint256(ref original) = self;
+        let mut ret = [0u64; 4];
+        let word_shift = shift / 64;
+        let bit_shift = shift % 64;
+        for i in 0..4 {
+            // Shift
+            if bit_shift < 64 && i + word_shift < 4 {
+                ret[i + word_shift] += original[i] << bit_shift;
+            }
+            // Carry
+            if bit_shift > 0 && i + word_shift + 1 < 4 {
+                ret[i + word_shift + 1] += original[i] >> (64 - bit_shift);
+            }
+        }
+        Uint256(ret)
+    }
+
+    /// Panic-free bitwise shift-right; yields `self >> mask(rhs)`, where `mask` removes any
+    /// high-order bits of `rhs` that would cause the shift to exceed the bitwidth of the type.
+    ///
+    /// Note that this is *not* the same as a rotate-right; the RHS of a wrapping shift-right is
+    /// restricted to the range of the type, rather than the bits shifted out of the LHS being
+    /// returned to the other end. We do not currently support `rotate_right`.
+    #[must_use = "this returns the result of the operation, without modifying the original"]
+    pub fn wrapping_shr(self, rhs: usize) -> Self {
+        let shift = if rhs > 255 { rhs & 0x000000ff } else { rhs };
+        let shift = shift as usize;
+
+        let Uint256(ref original) = self;
+        let mut ret = [0u64; 4];
+        let word_shift = shift / 64;
+        let bit_shift = shift % 64;
+        for i in word_shift..4 {
+            // Shift
+            ret[i - word_shift] += original[i] >> bit_shift;
+            // Carry
+            if bit_shift > 0 && i < 4 - 1 {
+                ret[i - word_shift] += original[i + 1] << (64 - bit_shift);
+            }
+        }
+        Uint256(ret)
+    }
+
+    #[cfg(test)]
+    fn bit(&self, index: usize) -> bool {
+        let &Uint256(ref arr) = self;
+        arr[index / 64] & (1 << (index % 64)) != 0
+    }
+
+    #[cfg(test)]
+    fn bit_slice(&self, start: usize, end: usize) -> Uint256 { (*self >> start).mask(end - start) }
+
+    #[cfg(test)]
+    fn mask(&self, n: usize) -> Uint256 {
+        let &Uint256(ref arr) = self;
+        let mut ret = [0; 4];
+        for i in 0..4 {
+            if n >= 0x40 * (i + 1) {
+                ret[i] = arr[i];
+            } else {
+                ret[i] = arr[i] & ((1 << (n - 0x40 * i)) - 1);
+                break;
+            }
+        }
+        Uint256(ret)
+    }
+}
+
+impl From<u64> for Uint256 {
+    fn from(x: u64) -> Uint256 { Uint256([x, 0, 0, 0]) }
+}
+
+impl PartialOrd for Uint256 {
+    fn partial_cmp(&self, rhs: &Uint256) -> Option<cmp::Ordering> { Some(self.cmp(rhs)) }
+}
+
+impl Ord for Uint256 {
+    fn cmp(&self, rhs: &Uint256) -> cmp::Ordering {
+        // We need to manually implement ordering because we use little-endian
+        // and the auto derive is a lexicographic ordering(i.e. memcmp)
+        // which with numbers is equivalent to big-endian
```

Comment:
I wonder if there is any advantage using little endian representation. There's at least the disadvantage that we can't derive `Ord`.

Reply:
I'm pretty confident with all the new/improved unit tests that I can trivially switch it to big-endian, @apoelstra do you remember why we used little-endian originally?

Reply:
Probably because it was easier for me to implement some algorithms with little endian, or maybe just easier to think about stuff .... I'm pretty sure it's fine to switch to BE internally.